src/main/java/ch/njol/skript/lang/function/ExprFunctionCall.java

```diff
@@ -13,31 +11,37 @@
 import org.bukkit.event.Event;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
+import org.skriptlang.skript.common.function.FunctionReference;
 import org.skriptlang.skript.lang.converter.Converters;
 
 import java.lang.reflect.Array;
-import java.util.*;
+import java.util.Arrays;
+import java.util.Map;
+import java.util.WeakHashMap;
 
 public class ExprFunctionCall<T> extends SimpleExpression<T> implements KeyProviderExpression<T> {
 
-	private final FunctionReference<?> function;
+	private final FunctionReference<?> reference;
 	private final Class<? extends T>[] returnTypes;
 	private final Class<T> returnType;
 	private final Map<Event, String[]> cache = new WeakHashMap<>();
 
-	public ExprFunctionCall(FunctionReference<T> function) {
-		this(function, function.returnTypes);
-	}
-
 	@SuppressWarnings("unchecked")
-	public ExprFunctionCall(FunctionReference<?> function, Class<? extends T>[] expectedReturnTypes) {
-		this.function = function;
-		Class<?> functionReturnType = function.getReturnType();
-		assert  functionReturnType != null;
-		if (CollectionUtils.containsSuperclass(expectedReturnTypes, functionReturnType)) {
+	public ExprFunctionCall(FunctionReference<?> reference, Class<? extends T>[] expectedReturnTypes) {
+		this.reference = reference;
+
+		Class<?> returnType;
+		Class<?> functionReturnType = reference.signature().returnType();
```

Comment:
might want to take the chance to future proof this with a possible return types-like system. supports future possible improvements like return type deduction or functions that return string or number but not player.

Optional.